packages/expo-updates/android/src/main/java/expo/modules/updates/codesigning/CodeSigningConfiguration.kt

```diff
@@ -0,0 +1,97 @@
+package expo.modules.updates.codesigning
+
+import android.util.Base64
+import android.util.Log
+import expo.modules.structuredheaders.BooleanItem
+import expo.modules.structuredheaders.Dictionary
+import expo.modules.structuredheaders.StringItem
+import java.security.Signature
+
+private const val TAG = "CodeSigning"
+
+/**
+ * Contains all information about code signing.
+ * @param embeddedCertificateString Implicitly trusted self-signed root certificate. May be the code signing certificate if only using embedded certificate (no chain).
+ * @param codeSigningMetadata Metadata about the code signing certificate. May be the root certificate or a leaf certificate.
+ * @param includeManifestResponseCertificateChain Should a certificate chain in the manifest response be evaluated when validating the signature. If true, ignores codeSigningMetadata when evaluating cert chain.
+ */
+class CodeSigningConfiguration(
+  private val embeddedCertificateString: String,
+  private val codeSigningMetadata: Map<String, String>?,
+  private val includeManifestResponseCertificateChain: Boolean,
+) {
+  private val algorithmFromMetadata: CodeSigningAlgorithm by lazy {
+    CodeSigningAlgorithm.parseFromString(
+      codeSigningMetadata?.get(
+        CODE_SIGNING_METADATA_ALGORITHM_KEY
+      )
+    )
+  }
+
+  private val keyIdFromMetadata: String by lazy {
+    codeSigningMetadata?.get(CODE_SIGNING_METADATA_KEY_ID_KEY) ?: CODE_SIGNING_METADATA_DEFAULT_KEY_ID
+  }
+
+  fun validateSignature(info: SignatureHeaderInfo, bodyBytes: ByteArray, manifestResponseCertificateChain: String?): Boolean {
+    val certificateChain = if (includeManifestResponseCertificateChain) {
+      CertificateChain(
+        separateCertificateChain(manifestResponseCertificateChain ?: "") + embeddedCertificateString
+      )
+    } else {
+      // check that the key used to sign the response is the same as the key in the code signing certificate
+      if (info.keyId != keyIdFromMetadata) {
+        throw Exception("Key with keyid=${info.keyId} from signature not found in client configuration")
+      }
+
+      // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
+      // algorithm specified in the configuration
+      if (info.algorithm != algorithmFromMetadata) {
+        Log.i(
+          TAG,
+          "Key with alg=${info.algorithm} from signature does not match client configuration algorithm, continuing"
+        )
+      }
+
+      CertificateChain(listOf(embeddedCertificateString))
+    }
+
+    return Signature.getInstance(
+      when (info.algorithm) {
```

Comment:
According to the comment, which I think is right, don't we want to use the algorithm specified by the client, not the server? This way if a malicious server says to use a weak (or no-op) algorithm, the client will ignore that and use its strong algorithm.

```suggestion
      // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
      // algorithm specified in the configuration
      if (info.algorithm != algorithmFromMetadata) {
        Log.i(
          TAG,
          "Key with alg=${info.algorithm} from signature does not match client configuration algorithm, continuing"
        )
      }

      CertificateChain(listOf(embeddedCertificateString))
    }

    return Signature.getInstance(
      when (algorithmFromMetadata) {
```

Reply:
The issue comes down to intermediate certificates (ones that are served in the manifest part). For expo go, we embed the root cert and the metadata is then for that cert (we probably shouldn't even specify the alg in the metadata for that cert). We download the intermediate cert but don't have a way of defining the metadata for that cert.

Maybe the right call is to specify that if a cert is downloaded as part of a chain and then served alongside the manifest it signed then saying the algorithm in the info should be used (what this change was supposed to do). Then, in the normal case (when the cert is not downloaded) we use the metadata from the root (the embedded metadata json).

Reply:
OK, I see. What I think we need to do is ignore `info.algorithm`, more or less, and not special-case chains. From https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12#section-2.5

> In order to verify a signature, a server MUST:
>
> ...
>
>  2.  The `algorithm`, `keyId`, and base 64 decoded `signature` listed
       in the Signature Parameters are then used to verify the
       authenticity of the digital signature.  **Note: The application
       verifying the signature MUST derive the digital signature
       algorithm from the metadata associated with the `keyId` and MUST
       NOT use the value of `algorithm` from the signed message.**

I think the attack they are defending against is one where a malicious signer sends a payload with:
```
keyId=something
signature=uryyb
algorithm=rot13
message=hello
```

and the verifier runs logic like:
```
if algorithm == rot13:
  return rot13(message) == signature
if algorithm === SHA256withRSA:
  ...
```

Instead, the verifier should do:
```
# ignore the signer's algorithm
key = get_key(keyId)
algorithm = get_algorithm(key)
if algorithm == ...
```

Reply:
Concretely speaking, we could have the implementation in expo-updates currently be:
```
fn get_algorithm(key):
  return SHA256withRSA. # this is all we support
```

and I believe that would be secure as long as the key is trusted (i.e. the cert chain is validated securely)